chacha20poly1305/src/lib.rs

```diff
@@ -0,0 +1,157 @@
+//! ChaCha20Poly1305 Authenticated Encryption with Additional Data Algorithm
+//! (RFC 8439)
+
+#![no_std]
+
+extern crate alloc;
+
+pub use aead;
+
+use aead::generic_array::typenum::{U0, U12, U16, U32};
+use aead::{generic_array::GenericArray, Aead, Error, NewAead};
+use alloc::vec::Vec;
+use chacha20::stream_cipher::{NewStreamCipher, SyncStreamCipher, SyncStreamCipherSeek};
+use chacha20::ChaCha20;
+use core::convert::TryInto;
+use poly1305::{Poly1305, Tag};
+use zeroize::{Zeroize, Zeroizing};
+
+/// ChaCha20Poly1305 AEAD
+#[derive(Clone)]
+pub struct ChaCha20Poly1305 {
+    /// Secret key
+    key: GenericArray<u8, U32>,
+}
+
+impl NewAead for ChaCha20Poly1305 {
+    type KeySize = U32;
+
+    fn new(key: GenericArray<u8, U32>) -> Self {
+        ChaCha20Poly1305 { key }
+    }
+}
+
+impl Aead for ChaCha20Poly1305 {
```

Comment:
You probably should implement `StatelessAead` instead. You'll get `Aead` via the [blanket impl](https://github.com/RustCrypto/traits/blob/master/aead/src/lib.rs#L87).

Reply:
Aah, my bad, I was worried about the `&mut self` that everything `salsa20-core`-based requires, but forgetting the fact I instantiate the cipher for every nonce. Will fix.